maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java

```diff
@@ -114,14 +123,42 @@ private CoreExtensionEntry createExtension( CoreExtension extension, List<Artifa
     {
         String realmId =
             "coreExtension>" + extension.getGroupId() + ":" + extension.getArtifactId() + ":" + extension.getVersion();
-        ClassRealm realm = classWorld.newRealm( realmId, null );
-        log.debug( "Populating class realm " + realm.getId() );
-        realm.setParentRealm( parentRealm );
+        final ClassRealm realm = classWorld.newRealm( realmId, null );
+        Set<String> providedArtifacts = Collections.emptySet();
+        String classLoadingStrategy = extension.getClassLoadingStrategy();
+        if ( STRATEGY_PARENT_FIRST.equals( classLoadingStrategy ) )
+        {
+            realm.importFrom( parentRealm, "" );
+        }
+        else if ( STRATEGY_PLUGIN.equals( classLoadingStrategy ) )
+        {
+            coreExports.getExportedPackages().forEach( ( p, cl ) -> realm.importFrom( cl, p ) );
+            providedArtifacts = coreExports.getExportedArtifacts();
+        }
+        else if ( STRATEGY_SELF_FIRST.equals( classLoadingStrategy ) )
+        {
+            realm.setParentRealm( parentRealm );
+        }
+        else
+        {
+            throw new IllegalArgumentException( "Unsupported class-loading strategy '"
+                    + classLoadingStrategy + "'. Supported values are: " + STRATEGY_PARENT_FIRST
+                    + ", " + STRATEGY_PLUGIN + " and " + STRATEGY_SELF_FIRST );
+        }
+        log.debug( "Populating class realm {}", realm.getId() );
         for ( Artifact artifact : artifacts )
         {
-            File file = artifact.getFile();
-            log.debug( "  Included " + file );
-            realm.addURL( file.toURI().toURL() );
+            String id = artifact.getGroupId() + ":" + artifact.getArtifactId();
+            if ( providedArtifacts.contains( id ) )
+            {
+                log.debug( "  Excluded {}", id );
+            }
+            else
+            {
+                File file = artifact.getFile();
+                log.debug( "  Included {}", file );
```

Comment:
This doesn't make sense. The exclude logs id and this one logs files. Inconsistent.
If you need to  print the file, I would recommend to print the id at debug level, but at trace level print both.

Reply:
I then propose the following code to avoid logging 2 lines in case trace is enabled:
```
                if ( log.isTraceEnabled() )
                {
                    log.debug( "  Included {} located at {}", id, file );
                }
                else
                {
                    log.debug( "  Included {}", id );
                }
```

Reply:
That's exactly what I have expected.

Reply:
> That's exactly what I have expected. Notice that the first log message must be `trace`. Likely just a typo.

Not a typo at all. If I want a trace statement, I'd have to use
```
log.debug( "  Included {}", id );
if ( log.isTraceEnabled() )
{
    log.trace( "            located at {}", file );
}
```
but then, the log output contains two lines at different levels which makes things less readable (if needed).  That's why I propose to log at level debug, but have a different output at debug/trace.  This is an usual pattern but I've used it several times in order to provide more information at certain level, see https://github.com/apache/maven-mvnd/blob/71631fd1919513aa8765008fb8c96fb7b4c82b6e/daemon/src/main/java/org/apache/maven/project/CachingProjectBuilder.java#L828-L834 for example
I think this makes the log output more readable while still conditioning the level of data that appears.
I haven't seen this pattern used in maven, hence the question which was very specific and without typo ;-)

Reply:
I consider this as an antipattern. Warn is warn, regardless which level is enabled. I am not really convinced by this. I would have expected this:
```
                if ( log.isTraceEnabled() )
                {
                    log.trace( "  Included {} located at {}", id, file );
                }
                else
                {
                    log.debug( "  Included {}", id );
                }
```

The message remains the same, but the granularity chagnes.

Reply:
Why do you I hve to enable debug to see more information with the same warn level?

Reply:
> I consider this as an antipattern. Warn is warn, regardless which level is enabled. 

I agree.

> I am not really convinced by this. 
> I would have expected this:
> 
> ```
>                 if ( log.isTraceEnabled() )
>                 {
>                     log.trace( "  Included {} located at {}", id, file );
>                 }
>                 else
>                 {
>                     log.debug( "  Included {}", id );
>                 }
> ```
> 
> The message remains the same, but the granularity changes.

This exactly contradicts your statement above which is "warn is warn".  Here, a _debug_ is logged either at _trace_ *or* _debug_ which is really wrong to me.

> Why do you I have to enable debug to see more information with the same warn level?
Same as here.  The level choosen in the `mvnd` example is _warn_ because that's what it is.  The exception stack trace however, is not necessary to understand the warning usually.  But in case you need to have more detailed information, you raise the log level to debug, and you have more information provided.  But that does not mean the _warning_ becomes a _trace_ as you said.

I think we have the exact same use case here.  The fact that an artifact is included or excluded is _debug_ level info.  If you need more, in this case the exact file location, you can raise the log level, but the semantic about the fact that the artifact is included or not does not change, nor should its log level.

If we can't agree on that solution, I'd go for the alternative using two statements, one at debug and one at trace level.
```
log.debug( "  Included {}", id );
log.trace( "          ->  {}", file );
```

Reply:
Let me think about this

Reply:
+1 for this pattern, I use it as well: emit (same) DEBUG message w/ less or more info... make possible grepping as well, etc.

Reply:
...still crunching... @cstamas Which pattern explicitly?

Reply:
The
```
if ( log.isTraceEnabled() )
{
    log.debug( "  Included {} located at {}", id, file );
}
else
{
    log.debug( "  Included {}", id );
}
```
one. These are two orthogonal things:
* message is always DEBUG
* level may be "tuned", ans based on set level, message either becomes invisible, visible, or more detailed.

Reply:
> > Warn is warn, regardless which level is enabled.
> 
> I agree.
> 
> > I am not really convinced by this.
> > I would have expected this:
> > ```
> >                 if ( log.isTraceEnabled() )
> >                 {
> >                     log.trace( "  Included {} located at {}", id, file );
> >                 }
> >                 else
> >                 {
> >                     log.debug( "  Included {}", id );
> >                 }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The message remains the same, but the granularity changes.
> 
> This exactly contradicts your statement above which is "warn is warn". Here, a _debug_ is logged either at _trace_ _or_ _debug_ which is really wrong to me.

You are right.

> > Why do you I have to enable debug to see more information with the same warn level?
> 
> The level chosen in the `mvnd` example is _warn_ because that's its semantic. The exception stack trace however, is not necessary to understand the warning usually. But in case you need to have more detailed information, you raise the log level to debug, and you have more information provided. But that does not mean the _warning_ becomes a _debug_ statement as you said.

Understand.

> I think we have the exact same use case here. The fact that an artifact is included or excluded is _debug_ level message. If you need more, in this case the exact file location, you can raise the log level, but the semantic about the fact that the artifact is included or not does not change, nor should its log level.
> 
> If we can't agree on that solution, I'd go for the alternative using two statements, one at debug and one at trace level.
> 
> ```
> log.debug( "  Included {}", id );
> log.trace( "          ->  {}", file );
> ```

That's not really better because it is split now in two lines which makes it even harder to grep.

Counter questions:
* Why not display the path always with debug?
* Why display the path with excludes, but not with inlcudes?

Reply:
> > If we can't agree on that solution, I'd go for the alternative using two statements, one at debug and one at trace level.
> > ```
> > log.debug( "  Included {}", id );
> > log.trace( "          ->  {}", file );
> > ```
> 
> That's not really better because it is split now in two lines which makes it even harder to grep.

Yes, I do agree and that's exactly why I did not want to use that.

> Counter questions:
> 
> * Why not display the path always with debug?

Yes, let's do that, it may be the simplest way forward.

> * Why display the path with excludes, but not with inlcudes?

I think it only makes sense when included for inspection when dealing with classloading problems.  When not included, the exact content does not matter so much imho.